K2Field.K2NE.ServiceBroker/ServiceObjects/ExcelServices/ExcelImportServiceSO.cs

```diff
@@ -0,0 +1,528 @@
+
+                                // Add the transaction ID value for identification purposes.
+                                if (sTransactionIDName != string.Empty && sTransactionIDValue != string.Empty)
+                                {
+                                    // replace spaces with underscores as SmartObject system names do that
+                                    newSmartObject.Properties[sTransactionIDName.Replace(" ", "_")].Value = sTransactionIDValue;
+                                }
+
+                                if (!string.IsNullOrEmpty(FQN))
+                                {
+                                    newSmartObject.Properties["FQN"].Value = FQN;
+                                }
+
+                                inputList.SmartObjectsList.Add(newSmartObject);
+                            }
+                            try
+                            {
+                                smoClientServer.ExecuteBulkScalar(soImport, inputList);
+                            }
+                            catch (SOC.SmartObjectException ex)
+                            {
+                                result = ex.Message + ex.StackTrace;
+                                foreach (SOC.SmartObjectExceptionData smOExBrokerData in ex.BrokerData)
+                                {
+                                    result += smOExBrokerData.Message + "  Before failure, " + result;
+                                }
+                                return result;
+                            }
+
+                            // Indicate the Transaction ID name and value in the results field
+                            if (sTransactionIDName != string.Empty && sTransactionIDValue != string.Empty)
+                            {
+                                result += ". " + sTransactionIDName + ":" + sTransactionIDValue;
+                            }
+                        }
+                        // free objects
+                        arrMatchingCols = null;
+                        soImport = null;
+                    }
+                    catch (ApplicationException ex)
+                    {
+                        result = "Unable to insert data into SmartObject: " + ex.Message;
+                        return result;
+                    }
+                }
+            }
+            catch (ApplicationException ex)
+            {
+                result = "Unable to connect to K2 server: " + ex.Message;
+            }
+            return result;
+        }
+
+
+        #region Helper functions
+
+        /// <summary>
+        ///  Read Data from selected excel file on client
+        /// </summary>
+        /// <param name="stream">The whole source Excel file passed in through memory stream, usually the server will do this.</param>
+        /// <returns>Successfully extracted rows.</returns>
+        private DataTable ReadExcelFile(MemoryStream stream, string sheetName)
```

Comment:
I would pass in the FileProperty.Content instead of the memorystream so all the reading is really handled by this method.

Reply:
Done.